### DISCUSSION TITLE
Flyway, 데이터베이스의 변경 이력 관리와 롤백의 필요성

### DISCUSSION THREAD

**Discussion by thornscript** [GU-99/grow-up-pms#351] [개발 토론]
# 개요
당신은 팀과 함께 새로운 온라인 쇼핑몰을 개발하고 있다. 초기에는 단순한 제품 목록과 사용자 관리 기능만 필요했기 때문에, 간단한 데이터베이스 스키마를 설정했다. 

하지만 프로젝트가 확장되면서, 카트 기능, 주문 관리, 결제 시스템 등 새로운 기능이 추가되면서 데이터베이스 스키마도 점점 복잡해지기 시작했다.

## 데이터베이스 스키마 관리의 복잡성 증가
시간이 지나면서, 개발자들은 여러 환경(개발, 테스트, 프로덕션)에서 작업을 하게 됐다. 개발자 A는 ‘주문 테이블’에 새로운 열을 추가했고, 개발자 B는 ‘사용자 테이블’에 인덱스를 추가했다. 각각의 변경 사항이 여러 환경에 일관되게 적용되어야 하는데, 수동으로 관리하기에는 점점 더 복잡해진다.

![image](https://github.com/user-attachments/assets/8d344dfa-c7b4-4dad-9833-886eddfbc01d)

만약, 누군가가 실수로 테스트 환경에서만 스키마 변경을 하고 운영 환경에 반영하지 않으면, 배포 시 오류가 발생할 수 있다. 예를 들어, 운영 환경에서 ‘주문 테이블’에 새 열이 없어서 애플리케이션이 정상적으로 동작하지 않을 수 있다.

## 변경 이력 관리와 롤백
프로젝트가 진행되는 동안, 잘못된 스키마 변경이 발생할 수 있다. 예를 들어, 특정 기능이 필요 없다고 판단해서 ‘할인 테이블’을 삭제했는데, 나중에 이 기능이 다시 필요해진 상황이다. 이때, 과거의 스키마 상태로 돌아가고 싶지만, 수동으로 작업한다면 상당한 시간이 걸릴 것이다. 그리고 실수로 데이터를 잃을 위험도 크다.

## 팀 협업의 효율성 향상 필요성
프로젝트가 확장됨에 따라 팀 내 협업의 중요성이 더욱 커졌다. 개발자들은 각자 다른 기능을 개발하면서 데이터베이스 스키마를 변경하게 되었다. 예를 들어, 개발자 C는 새로운 '리뷰 시스템'을 구현하면서 관련 테이블을 추가해야 했고, 개발자 D는 기존 '제품 테이블'의 구조를 최적화하고 있었다.

이러한 변경사항을 효과적으로 공유하고 관리하는 것이 중요해졌지만, 현재의 방식으로는 한계가 있었다. 개발자들은 자신의 변경사항을 팀에 알리기 위해 이메일이나 메신저를 사용했지만, 이는 종종 누락되거나 잊혀지곤 했다. 또한, 데이터베이스 변경사항을 코드 리뷰 프로세스에 포함시키기 어려워, 잠재적인 문제를 사전에 발견하고 해결하는 데 어려움을 겪었다.

팀은 데이터베이스 변경사항을 코드처럼 관리하고, 버전 관리 시스템을 통해 쉽게 공유할 수 있는 방법이 필요하다고 느꼈다. 이를 통해 코드 리뷰 프로세스에서 데이터베이스 변경사항도 함께 검토할 수 있게 되어, 협업의 효율성을 높이고 잠재적인 문제를 사전에 방지할 수 있을 것이라 기대했다.

## 데이터 마이그레이션의 자동화 필요성
온라인 쇼핑몰의 성장에 따라 기존 데이터 구조를 변경해야 하는 상황이 빈번히 발생했다. 예를 들어, 사용자 주소 정보를 더 세분화하여 저장하기 위해 '주소 테이블'의 구조를 변경해야 했다. 이는 단순한 스키마 변경뿐만 아니라 기존 데이터를 새로운 구조에 맞게 변환해야 하는 복잡한 작업이었다.

개발팀은 이러한 데이터 마이그레이션 작업을 수동으로 수행하고 있었는데, 이는 여러 문제를 야기했다. 작업 과정에서 실수가 발생하기 쉬웠고, 대량의 데이터를 다루다 보니 시간도 많이 소요되었다. 또한, 개발, 테스트, 운영 환경 각각에서 동일한 마이그레이션 작업을 반복해야 했기 때문에 비효율적이었다.

팀은 이러한 데이터 마이그레이션 로직을 자동화하고 버전 관리할 수 있는 방법이 필요하다고 느꼈다. 이를 통해 복잡한 데이터 변환 작업을 일관되게 실행하고, 필요시 여러 환경에서 쉽게 재현할 수 있기를 원했다. 또한, 문제가 발생했을 때 이전 상태로 안전하게 롤백할 수 있는 메커니즘도 필요하다고 판단했다.

# 정의
`Flyway`는 데이터베이스 형상 관리 도구이다. 데이터베이스의 스키마나 데이터를 관리할 때, 버전 관리나 마이그레이션 등의 일을 자동화할 수 있다. 따라서 개발자의 실수를 줄일 수 있으며, 여러 개발자들이 서로 협업을 할 때 데이터베이스의 상태를 일관되게 유지할 수 있다는 장점이 있다.

# 개념 살펴보기
## 마이그레이션
Flyway에서는 DB에 대한 모든 변경 사항을 **마이그레이션**이라고 한다. 마이그레이션에는 `버전(versioned)`과 `반복(repeatable)` 두 가지 타입이 있다.

### 버전 마이그레이션
버전 마이그레이션은 버전, 설명, 체크섬을 필요로 한다. 버전은 고유해야 하며, 설명은 각 마이그레이션이 무엇을 하는지 알려주기 위해 작성하는 것으로 순전히 정보 제공용이라고 할 수 있다. 체크섬은 실수로 인한 변경 사항을 감지하기 위해서 존재한다. 버전 마이그레이션은 가장 일반적인 마이그레이션 타입으로, 버전 순서대로 한 번만 적용된다.

버전 마이그레이션은 일반적으로 아래와 같은 경우에 사용된다.

- 테이블/인덱스/외래 키/열거형/UDT 등의 생성/변경/삭제
- 참조 데이터 업데이트
- 사용자 데이터 수정

```sql
CREATE TABLE car (
    id INT NOT NULL PRIMARY KEY,
    license_plate VARCHAR NOT NULL,
    color VARCHAR NOT NULL
);

ALTER TABLE owner ADD driver_license_id VARCHAR;

INSERT INTO brand (name) VALUES ('DeLorean');
```

### 반복 마이그레이션
반복 마이그레이션은 설명, 체크섬을 필요로 하지만 버전은 따로 쓰지 않아도 된다. 한 번만 실행되는 대신, 체크섬이 변경될 때마다 (재)적용된다.

마이그레이션을 실행할 때, 반복 마이그레이션은 항상 마지막에 적용되며, 모든 대기 중인 버전 관리 마이그레이션이 실행된 후에 적용된다.

### 취소 마이그레이션
Flyway Teams 유료 버전에서만 사용 가능한 기능이다. 취소 마이그레이션은 버전 마이그레이션의 반대다. 취소 마이그레이션은 동일한 버전의 버전 마이그레이션 변경 사항을 되돌리는 역할을 하며, 아래와 같이 마이그레이션 작업을 반대 방향으로 스키마를 변경한다.

```sql
DELETE FROM brand WHERE name='DeLorean';

ALTER TABLE owner DROP driver_license_id;

DROP TABLE car;
```

프로덕션 환경에서는 신중하게 사용해야 하며, 롤백이 실제로 수행되기 전에는 항상 백업을 만들고, 롤백 후에는 데이터의 일관성을 확인하는 것이 좋다.

## 메타데이터 테이블
Flyway는 데이터베이스 스키마의 상태를 보존하기 위해서 모든 스키마 마이그레이션에 대해 메타데이터를 `flyway_schema_history`라는 자체 테이블에 생성하고 저장한다.

![image](https://github.com/user-attachments/assets/e065ca87-063a-44d9-ac0c-74316a28eaf4)

테이블에는 어떤 순서로, 어떤 스크립트가 언제 적용되었는지 기록된다. 그리고 Flyway는 데이터베이스에 스크립트를 적용한 후 스크립트의 불변성을 유지하기 위해서 각 마이그레이션 스크립트의 체크섬을 저장하고 있다.

스크립트를 적용할 때 Flyway는 이미 적용된 스크립트가 있는 경우에 저장된 체크섬을 마이그레이션 디렉토리에 있는 모든 기존 스크립트에 대한 체크섬을 생성하여 이를 비교한다. 하나라도 체크섬이 일치하지 않으면 마이그레이션이 중단된다. 그렇지 않으면 Flyway는 어떤 스크립트가 새로운지 확인하고 버전 번호별로 정렬해서 적용한다.

# 빠른 시작
## 의존성 추가
```groovy
dependencies {
	implementation 'org.flywaydb:flyway-core'
}
```

## 설정 파일 추가
```yaml
spring:
  flyway:
    enabled: true
    # Flyway가 마이그레이션 스크립트를 찾는 위치
    locations: classpath:db/migration/structure, classpath:db/migration/data
    # 이 기능은 이미 DB에 데이터가 있지만 스키마 히스토리 테이블이 없는 DB에 대해 마이그레이션을 실행할 때 기준점(baseline)을 설정할지 여부를 결정한다.
    # 이 기능을 켜면, 마이그레이션을 실행하기 전에 baselineVersion으로 지정된 버전을 기준점으로 삼게 된다. 그 후에는 baselineVersion보다 높은 버전의 마이그레이션만 적용된다.
    # 이 기능은 이미 운영 중인 DB에 Flyway를 처음 도입할 때 유용하다.
    baseline-on-migrate: true
```

## 데이터베이스 마이그레이션 스크립트 생성
마이그레이션 스크립트를 생성할 때는 `V<버전>__<설명>sql`, `R__<설명>.sql`과 같이 Flyway에서 정한 포맷을 따라야 한다.

- **접두사(prefix):** V, U, R 중 하나를 입력해야 하는데 V는 버전(version), U는 취소(undo), R은 반복(repeatable)이다.
- **버전:** 버전 정보이며 정수, 소수, 날짜 등이 가능하다. 추가할 때는 항상 최신 마이그레이션 스크립트보다 큰 숫자로 버전을 설정해야 한다.
- **구분자(separator):** 언더바(`_`) 2개를 이어서 작성한다.
- **설명(description):** 스크립트에 대한 추가적인 설명이다.

>[!WARNING]
>스크립트 작성 시에 아래의 규칙을 따라야 한다. 그렇지 않으면 마이그레이션 도중에 오류가 발생한다.
>- Flyway 마이그레이션 스크립트 파일명에는 버전 번호가 중복되지 않아야 한다. (예: `V1__init.sql`, `V1__indexes.sql`는 버전 번호 1이 중복으로 사용됨)
>- 마이그레이션이 적용된 후에는 스크립트의 내용을 변경하면 안된다.

## 참고
- [Flyway Documentation - Migrations](https://documentation.red-gate.com/fd/migrations-184127470.html)

**Comment by wlshooo** [comment]
데이터베이스 현상관리는 사용해본 적은 없지만 러닝커브가 낮으면 미래지향적으로, 적용해두었을 때 관리적인 측면에서 이점이 많은 거 같습니다!

**Reply by thornscript** [reply]
동일한 문제가 반복되면 다음 백엔드 회의 때 같이 모여서 의논해보는 것도 좋을 것 같습니다 :smile:
